modules/internal/ExternalArray.chpl

```diff
@@ -0,0 +1,284 @@
+      var r = inds(1);
+      if (r.low != 0) {
+        halt("external arrays always have a lower bound of 0");
+      }
+      var newdom = new unmanaged ArrayViewExternDom(idxType,
+                                                    r.size,
+                                                    _to_unmanaged(this));
+      return newdom;
+    }
+
+    proc dsiClone() {
+      return new unmanaged ArrayViewExternDist();
+    }
+  }
+
+  class ArrayViewExternDom: BaseRectangularDom {
+    const size: uint; // We don't need a lower bound, it will always be zero
+
+    const dist;
+
+    proc init(type idxType, size, dist) {
+      super.init(1, idxType, false);
+      this.size = size: uint;
+      this.dist = dist;
+    }
+
+    proc dsiBuildArray(type eltType) {
+      var arr = new unmanaged ArrayViewExternArr(eltType,
+                                                 _to_unmanaged(this),
+                                                 c_malloc(eltType, this.size),
+                                                 true);
+      return arr;
+    }
+
+    proc dsiBuildArrayWith(type eltType, data:c_ptr(eltType), allocSize: int) {
+      var arr = new unmanaged ArrayViewExternArr(eltType,
+                                                 _to_unmanaged(this),
+                                                 data,
+                                                 false);
+      return arr;
+    }
+
+    proc domRange return 0..#size;
+
+    proc dsiGetIndices() return (domRange,);
+
+    proc dsiSetIndices(x) {
+      halt("Can't change the indices of an external array");
+    }
+
+    iter these() {
+      for i in domRange do
+        yield i;
+    }
+
+    iter these(param tag: iterKind) where tag == iterKind.standalone {
+      forall i in domRange do
+        yield i;
+    }
+
+    iter these(param tag: iterKind) where tag == iterKind.leader {
+      for followThis in domRange {
+        yield followThis;
+      }
+    }
+
+    iter these(param tag: iterKind, followThis)
+      where tag == iterKind.follower {
+      for i in domRange do
+        yield i;
+    }
+
+    proc dsiMyDist() {
+      return dist;
+    }
+
+    proc dsiMember(ind: rank*idxType) {
+      if (ind(1) < size && ind(1) >= 0) {
+        return true;
+      } else {
+        return false;
+      }
+    }
+
+    proc dsiDim(d: int) {
+      if (d != rank) {
+        halt("domains over external arrays have only one dimension");
+      }
+      return domRange;
+    }
+
+    // Necessary?
+    proc dsiDim(param d: int) {
+      if (d != rank) {
+        halt("domains over external arrays have only one dimension");
+      }
+      return dsiGetIndices();
+    }
+
+    proc dsiAssignDomain(rhs: domain, lhsPrivate: bool) {
+      chpl_assignDomainWithGetSetIndices(this, rhs);
+    }
+
+    // What about _getActualDom?  dsiDestroyDom?
+
+    // Prolly want the privatization stuff eventually, but I don't need it right
+    // now.
+  }
+
+  class ArrayViewExternArr: BaseArr {
```

Comment:
If it has the option of "owning" the data, I wouldn't call it ArrayViewExternArr. We could call it just ExternArr. Or we could make it not allowed to "own" the data.